This project does exploraation of ebay car sales 

In [1]:
import pandas as pd
import numpy as np

In [4]:
autos = pd.read_csv('autos.csv' , encoding='Latin-1')

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

pretty small dataset, only has 7.6mb in memory. Describes cars and its features. It has 50k entries of diff cars sold. Need to convert some numeric values to numbers from objects. Some columns have lots of missing values.

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [27]:
copied = autos.columns.tolist().copy()

In [28]:
copied

['dateCrawled',
 'name',
 'seller',
 'offerType',
 'price',
 'abtest',
 'vehicleType',
 'yearOfRegistration',
 'gearbox',
 'powerPS',
 'model',
 'odometer',
 'monthOfRegistration',
 'fuelType',
 'brand',
 'notRepairedDamage',
 'dateCreated',
 'nrOfPictures',
 'postalCode',
 'lastSeen']

In [30]:
copied = ['date_crawled',
 'name',
 'seller',
 'offer_type',
 'price',
 'abtest',
 'vehicle_type',
 'registration_year',
 'gearbox',
 'power_ps',
 'model',
 'odometer',
 'registration_month',
 'fuel_type',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'nr_of_pictures',
 'postal_code',
 'last_seen']

In [31]:
autos.columns = copied

In [32]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

snake case is for readability.

In [33]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

registration year cant' be 1000 or 9999, nr of pics needs more closer look
unrepaired damage is 'nein' must be 'no'. price and odometer must be numeric.

In [36]:
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(float)

In [37]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)

In [38]:
autos.rename({'odometer': 'odometer_km'}, inplace=True, axis=1)

In [39]:
autos['price'].shape

(50000,)

In [40]:
autos['odometer_km'].shape

(50000,)

In [41]:
autos['price'].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [43]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [44]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [48]:
autos['date_crawled'].str[:7].value_counts(normalize=True)

2016-03    0.8379
2016-04    0.1621
Name: date_crawled, dtype: float64

In [61]:
autos['ad_created'].sort_index().str[:7].value_counts(normalize=False)

2016-03    41866
2016-04     8053
2016-02       63
2016-01       12
2015-12        2
2015-06        1
2015-11        1
2015-08        1
2015-09        1
Name: ad_created, dtype: int64

In [60]:
autos['last_seen'].sort_index().str[:7].value_counts(normalize=False)

2016-04    28709
2016-03    21291
Name: last_seen, dtype: int64

In [59]:
autos['registration_year'].sort_index().value_counts().head(3)

2000    3354
2005    3015
1999    3000
Name: registration_year, dtype: int64

top 3 registration years for cars are 2000, 2005,1999.

last seen cars are both in 2016

most ads are created in the first 4 months of 2016

highest and lowest value for registration year is 1950 and 2019 as when the cars were created and the current year

In [78]:
autos[autos['registration_year'] == 9999]

date_crawled                                               name  \
8012   2016-03-23 16:43:29                      Opel_GT_Karosserie_mit_Brief!   
14341  2016-03-23 01:36:20                                  Hole_kostenlos_ab   
33950  2016-03-23 21:52:25  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...   
38076  2016-04-04 22:54:47                                 Mercedes_Benz_A180   

       seller offer_type  price abtest vehicle_type  registration_year  \
8012   privat    Angebot    700   test          NaN               9999   
14341  privat    Angebot      0   test          NaN               9999   
33950  privat    Angebot   7999   test          NaN               9999   
38076  privat    Angebot  18000   test          NaN               9999   

      gearbox  power_ps     model  odometer_km  registration_month fuel_type  \
8012      NaN         0    andere      10000.0                   0       NaN   
14341     NaN         0       NaN      10000.0                   0       NaN   
33950     NaN         0    kaefer      10000.0                   0       NaN   
38076     NaN         0  a_klasse      10000.0                   0    benzin   

               brand unrepaired_damage           ad_created  nr_of_pictures  \
8012            opel               NaN  2016-03-23 00:00:00               0   
14341            bmw               NaN  2016-03-23 00:00:00               0   
33950     volkswagen               NaN  2016-03-23 00:00:00               0   
38076  mercedes_benz               NaN  2016-04-04 00:00:00               0   

       postal_code            last_seen  
8012         21769  2016-04-05 20:16:15  
14341        32689  2016-03-23 08:47:00  
33950        47638  2016-04-06 03:46:40  
38076        51379  2016-04-07 02:44:52

In [79]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [81]:
autos.iloc[[8012,14341,33950,38076], 7] =2019

In [82]:
autos['registration_year'].max()

9996

hmm seems like after i removed 9999 year, i still have some bad years like 9996, need to check what are the max years!

In [88]:
autos['registration_year'].value_counts().sort_index(ascending=False)

9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       7
2018     492
2017    1453
2016    1316
2015     399
2014     666
2013     806
2012    1323
2011    1634
2010    1597
2009    2098
2008    2231
2007    2304
2006    2708
2005    3015
2004    2737
2003    2727
2002    2533
2001    2703
2000    3354
        ... 
1963       9
1962       4
1961       6
1960      34
1959       7
1958       4
1957       2
1956       5
1955       2
1954       2
1953       1
1952       1
1951       2
1950       3
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       9
1800       2
1500       1
1111       1
1001       1
1000       1
Name: registration_year, Length: 96, dtype: int64

In [ ]:
When removing outliers, we can 
do df[(df["col"] > x ) & (df["col"] < y )], but
it's more readable to use df[df["col"].between(x,y)]


In [96]:
bad_years = autos[autos['registration_year'] > 2019]['registration_year']

In [109]:
bad_years

453      4500
4164     5000
4549     4100
6308     9996
8360     6200
13559    9000
22799    5000
24519    5000
25003    8888
27578    2800
27618    5911
42079    4800
49153    5000
49910    9000
Name: registration_year, dtype: int64

In [111]:
bad_index = bad_years.tolist()

In [112]:
bad_index

[4500,
 5000,
 4100,
 9996,
 6200,
 9000,
 5000,
 5000,
 8888,
 2800,
 5911,
 4800,
 5000,
 9000]

In [106]:
min_bad_years = autos[autos['registration_year'] <1950 ]['registration_year'].tolist()
#min_bad_index = min_bad_years.tolist()

In [107]:
min_bad_years

[1934,
 1934,
 1910,
 1800,
 1948,
 1931,
 1943,
 1941,
 1910,
 1927,
 1937,
 1929,
 1000,
 1910,
 1937,
 1111,
 1939,
 1941,
 1938,
 1937,
 1910,
 1910,
 1800,
 1910,
 1500,
 1937,
 1910,
 1910,
 1910,
 1001]

In [117]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [123]:
x = autos['brand'].value_counts().sort_values(ascending=False)

In [130]:
x = pd.DataFrame(x)

In [132]:
x = x.head(3)

In [133]:
x

brand
volkswagen  10687
opel         5461
bmw          5429

дывоадывдаоыдоаоыдвдыодоыдоаыодо


ыдвоадыоадыдаоыдовадыовдоафждводао209о203о
230ок230ао2-а0о3а-03оа20ао
ывоа
ывао-23оа02оавыоыдао


In [135]:
autos['mean_price'] = 0

In [136]:
autos['mean_mileage'] = 0

In [137]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   
2    Angebot   8990     test    limousine               2009    manuell   

   power_ps     ...      registration_month  fuel_type       brand  \
0       158     ...                       3        lpg     peugeot   
1       286     ...                       6     benzin         bmw   
2       102     ...                       7     benzin  volkswagen   

  unrepaired_damage           ad_created nr_of_pictures postal_code  \
0              nein  2016-03-26 00:00:00              0       79588   
1              nein  2016-04-04 00:00:00              0       71034   
2              nein  2016-03-26 00:00:00              0       35394   

             last_seen  mean_price mean_mileage  
0  2016-04-06 06:45:54           0            0  
1  2016-04-06 14:45:08           0            0  
2  2016-04-06 20:15:37           0            0  

[3 rows x 22 columns]

In [138]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [139]:
autos[autos['seller'] == 'privat'] = 'private'

In [140]:
autos['seller'].value_counts()

private       49999
gewerblich        1
Name: seller, dtype: int64

In [141]:
autos[autos['seller'] == 'gewerblich'] = 'public'

In [142]:
autos

date_crawled     name   seller offer_type    price   abtest  \
0          private  private  private    private  private  private   
1          private  private  private    private  private  private   
2          private  private  private    private  private  private   
3          private  private  private    private  private  private   
4          private  private  private    private  private  private   
5          private  private  private    private  private  private   
6          private  private  private    private  private  private   
7          private  private  private    private  private  private   
8          private  private  private    private  private  private   
9          private  private  private    private  private  private   
10         private  private  private    private  private  private   
11         private  private  private    private  private  private   
12         private  private  private    private  private  private   
13         private  private  private    private  private  private   
14         private  private  private    private  private  private   
15         private  private  private    private  private  private   
16         private  private  private    private  private  private   
17         private  private  private    private  private  private   
18         private  private  private    private  private  private   
19         private  private  private    private  private  private   
20         private  private  private    private  private  private   
21         private  private  private    private  private  private   
22         private  private  private    private  private  private   
23         private  private  private    private  private  private   
24         private  private  private    private  private  private   
25         private  private  private    private  private  private   
26         private  private  private    private  private  private   
27         private  private  private    private  private  private   
28         private  private  private    private  private  private   
29         private  private  private    private  private  private   
...            ...      ...      ...        ...      ...      ...   
49970      private  private  private    private  private  private   
49971      private  private  private    private  private  private   
49972      private  private  private    private  private  private   
49973      private  private  private    private  private  private   
49974      private  private  private    private  private  private   
49975      private  private  private    private  private  private   
49976      private  private  private    private  private  private   
49977      private  private  private    private  private  private   
49978      private  private  private    private  private  private   
49979      private  private  private    private  private  private   
49980      private  private  private    private  private  private   
49981      private  private  private    private  private  private   
49982      private  private  private    private  private  private   
49983      private  private  private    private  private  private   
49984      private  private  private    private  private  private   
49985      private  private  private    private  private  private   
49986      private  private  private    private  private  private   
49987      private  private  private    private  private  private   
49988      private  private  private    private  private  private   
49989      private  private  private    private  private  private   
49990      private  private  private    private  private  private   
49991      private  private  private    private  private  private   
49992      private  private  private    private  private  private   
49993      private  private  private    private  private  private   
49994      private  private  private    private  private  private   
49995      private  private  private    private  private  private   
4999